### ISSUE TITLE
Need assistance with Windows 10

### ISSUE THREAD

**Issue by TheAJArts** [flareteam/flare-engine#1723]
Hello, 

(Windows 10 64-Bit)

I have (or so I think) all the necessary SDL and C++ files installed.

 

Cygwin terminal is running inside of the correct folder.

I feel like I'm so close to getting the project to build... When I run the make command, I get this:

`$ make
Building file: ../src/api/MapSaver.cpp
Invoking: Cross G++ Compiler
g++ -O0 -g3 -Wall -c -fmessage-length=0 -MMD -MP -MF"src/api/MapSaver.d" -MT"src/api/MapSaver.o" -o "src/api/MapSaver.o" "../src/api/MapSaver.cpp"
../src/api/MapSaver.cpp: In constructor ‘MapSaver::MapSaver(Map*)’:
../src/api/MapSaver.cpp:23:23: error: ‘EC_TOOLTIP’ was not declared in this scope
  EVENT_COMPONENT_NAME[EC_TOOLTIP] = "tooltip";
                       ^~~~~~~~~~
../src/api/MapSaver.cpp:24:23: error: ‘EC_POWER_PATH’ was not declared in this scope
  EVENT_COMPONENT_NAME[EC_POWER_PATH] = "power_path";`
 

 

+ A list of errors for all other occurences.
Now, when I go to the file, I get this error for all of them: 
"Symbol 'EC_TOOLTIP' could not be resolved" Any help would be appreciated

**Comment by dorkster** [comment]
Those files aren't required for building Flare normally (the CMake script excludes them). I've pushed an [update](https://github.com/flareteam/flare-engine/commit/d7435bd4efd5e5ee8b4b423cae27c698912962bb) that wraps these files with an `#ifdef`, so they shouldn't be accidentally included when building without CMake now. Let me know if that works.

**Comment by TheAJArts** [comment]
Thank you! I'm testing now. I will let you know in a minute.

Also, should I be worried about the missing EC_SCRIPT? 
I see it in the fix, but it's not in the MapSaver.cpp on mine

**Comment by TheAJArts** [comment]
Nevermind, I see that it is being added but not being removed, my mistake.

**Comment by TheAJArts** [comment]
It's being built! Thank you very much.

**Comment by TheAJArts** [comment]
I'm still confused.. I believe it finished, but I'm not sure what file I'm supposed to execute.

**Comment by TheAJArts** [comment]
When I run the make command, I get:

make: 'src/api/MapSaver.o' is up to date.

**Comment by dorkster** [comment]
I'm confused. You're using `make`, but if `MapSaver.cpp` is included then the Makefile wasn't created with CMake. So where is the Makefile you're using?

Maybe the [instructions for building without CMake](https://github.com/flareteam/flare-engine/blob/master/INSTALL.engine.md#building-with-g) could be helpful if you haven't looked at them already.

**Comment by TheAJArts** [comment]
Okay so,

fatal error: SDL.h: No such file or directory
 #include <SDL.h>


But I moved the SDL.h to the src directory of the engine.

**Comment by TheAJArts** [comment]
Nevermind, I just moved the SDL folder to the minGW folders and it is building now.

**Comment by TheAJArts** [comment]
I've been at this all night.. 

I know nothing about building with MinGW.

When I try to build, with:

> g++ -I C:\MinGW\include\SDL src/*.cpp -o flare.exe -lmingw32 -lSDLmain -lSDL2 -lSDL2_image -lSDL2_mixer -lSDL2_ttf


> 

I get: 

> 
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lmingw32
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDLmain
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL2_image
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL2_mixer
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL2_ttf
collect2: error: ld returned 1 exit status
> 

I understand that libraries aren't in the correct area, but I have NO idea where I'm supposed to place them. I've tried every folder in the MinGW installation.

It's giving me the usr/lib/gcc folder, but I cannot find it for the life of me to place the files...

**Comment by dorkster** [comment]
In the SDL mingw archive, there should be two folders for i686 and x86_64. Inside each of those is the following:

```
bin/
include/
lib/
share/
```

The base mingw folder should contain the same folders. In my case, this is `C:\Program Files (x86)\CodeBlocks\MinGW`. So copying those folders with their contents into your mingw base directory should do the trick.

**Comment by TheAJArts** [comment]
Sorry for the incompetence, I am trying my best to learn.

I have done as you instructed and the terminal has found the libs, but now I'm getting a list of these:

> C:/MinGW/include/SDL/SDL_endian.h:213:1: note: ‘float SDL_SwapFloat(float)’ previously defined here
>  SDL_SwapFloat(float x)
>  ^~~~~~~~~~~~~
> In file included from /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/include/SDL_thread.h:35:0,
>                  from /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/include/SDL_audio.h:35,
>                  from /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/include/SDL_mixer.h:27,
>                  from src/SDLSoundManager.h:29,
>                  from src/SDLSoundManager.cpp:31:
> /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/include/SDL_atomic.h: At global scope:
> /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/include/SDL_atomic.h:189:31: error: conflicting declaration ‘typedef struct SDL_atomic_t SDL_atomic_t’
>  typedef struct { int value; } SDL_atomic_t;
>

**Comment by dorkster** [comment]
If SDL was installed as part of Cygwin, I would try removing it.

Alternatively, removing `-I C:\MinGW\include\SDL` so that Cygwin's include file are used might work.

**Comment by TheAJArts** [comment]
Here is the response to your alternative:

```
$ g++ src/*.cpp -o flare.exe -lmingw32 -lSDL_main -lSDL2 -lSDL_image -lSDL_mixer -lSDL_ttf      /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lmingw32
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL_main
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL_image
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL_mixer
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL_ttf
collect2: error: ld returned 1 exit status

```

Here is the directory that cygwin is pointing to:

```
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0
$ ls
begin_code.h       SDL_clipboard.h              SDL_render.h
cc1.exe            SDL_config.h                 SDL_revision.h
cc1plus.exe        SDL_cpuinfo.h                SDL_rwops.h
close_code.h       SDL_egl.h                    SDL_scancode.h
collect2.exe       SDL_endian.h                 SDL_sensor.h
crtbegin.o         SDL_error.h                  SDL_shape.h
crtbeginS.o        SDL_events.h                 SDL_stdinc.h
crtend.o           SDL_filesystem.h             SDL_surface.h
crtfastmath.o      SDL_gamecontroller.h         SDL_system.h
cyglto_plugin.dll  SDL_gesture.h                SDL_syswm.h
include            SDL_haptic.h                 SDL_test.h
install-tools      SDL_hints.h                  SDL_test_assert.h
libatomic.a        SDL_image.h                  SDL_test_common.h
libatomic.dll.a    SDL_joystick.h               SDL_test_compare.h
libgcc.a           SDL_keyboard.h               SDL_test_crc32.h
libgcc_eh.a        SDL_keycode.h                SDL_test_font.h
libgcc_s.dll.a     SDL_loadso.h                 SDL_test_fuzzer.h
libgcov.a          SDL_log.h                    SDL_test_harness.h
libgomp.a          SDL_main.h                   SDL_test_images.h
libgomp.dll.a      SDL_messagebox.h             SDL_test_log.h
libgomp.spec       SDL_mixer.h                  SDL_test_md5.h
libitm.a           SDL_mouse.h                  SDL_test_memory.h
libitm.spec        SDL_mutex.h                  SDL_test_random.h
libquadmath.a      SDL_name.h                   SDL_thread.h
libquadmath.dll.a  SDL_opengl.h                 SDL_timer.h
libstdc++.a        SDL_opengl_glext.h           SDL_touch.h
libstdc++.dll.a    SDL_opengles.h               SDL_ttf.h
libstdc++fs.a      SDL_opengles2.h              SDL_types.h
libsupc++.a        SDL_opengles2_gl2.h          SDL_version.h
lto1.exe           SDL_opengles2_gl2ext.h       SDL_video.h
lto-wrapper.exe    SDL_opengles2_gl2platform.h  SDL_vulkan.h
SDL.h              SDL_opengles2_khrplatform.h  SDL2
SDL_assert.h       SDL_pixels.h                 SDL2.dll
SDL_atomic.h       SDL_platform.h               SDL2.lib
SDL_audio.h        SDL_power.h                  SDL2main.lib
SDL_bits.h         SDL_quit.h                   SDL2test.lib
SDL_blendmode.h    SDL_rect.h

```

**Comment by dorkster** [comment]
Okay, this is becoming clearer. Notice that `-lSDL2` worked? That's because `SDL2.lib` is in `/usr/lib/gcc/x86_64-pc-cygwin/7.4.0`.

You could also get `SDL2main.lib`, but you need to use `-lSDL2main` instead of `-lSDL_main` (note the '2' and lack of underscore).

The `SDL2_mixer`, `SDL2_image`, and `SDL2_ttf` libs appear to be missing. I would try to install them via Cygwin if possible.

**Comment by TheAJArts** [comment]
Yes, I did notice that. Okay, I see now. Yes, I have all the files for them as well. Give me a few minutes and I let you know the results.

**Comment by dorkster** [comment]
You'll probably have to remove `-lmingw32` as well because you're not building with MinGW, but rather Cygwin's GCC.

**Comment by TheAJArts** [comment]
Yeah, I did that after I posted the long ls list, thank you

**Comment by TheAJArts** [comment]
Progress update: 
Moved the main lib to the directory anddddd

```
$ g++ src/*.cpp -o flare.exe -lSDL2main -lSDL2 -lSDL2_image -lSDL2_mixer -lSDL2_ttf             /usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL2_image
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL2_mixer
/usr/lib/gcc/x86_64-pc-cygwin/7.4.0/../../../../x86_64-pc-cygwin/bin/ld: cannot find -lSDL2_ttf
collect2: error: ld returned 1 exit status


Getting there, let me just locate the rest of the libs and I'll post a final result.

```

**Comment by TheAJArts** [comment]
Okay, so i ran:

`g++ src/*.cpp -o flare.exe -lSDL2main -lSDL2 -lSDL2_image -lSDL2_mixer -lSDL2_ttf  `


and the flare.exe was built!
But, when I ran it, I was getting a missing .dll for all of the SDL(SDL2, image, mixer, and ttf) files.

**Comment by dorkster** [comment]
The DLL files should be in the same directory where the `flare.exe` is.

**Comment by TheAJArts** [comment]
AH, okay!

**Comment by TheAJArts** [comment]
All of the required libraries are in the flare.exe directory.

The file opens up cmd for a split second then creates a file flare.exe.stackdump

inside:
```
Exception: STATUS_ACCESS_VIOLATION at rip=003FEE2BE03
rax=00000000FFFFC9A0 rbx=0000000600028420 rcx=000000010063D988
rdx=00000000FFFFC9A0 rsi=0000000000000042 rdi=00000000006A23F6
r8 =0000000000000000 r9 =00000006000859C8 r10=0000000100000000
r11=00000003FEE885FC r12=000000000000005A r13=0000000000000000
r14=0000000000000000 r15=00000000FFFFCC6F
```

**Comment by dorkster** [comment]
Hmm, that's not very helpful. From the command line, try running `flare.exe --version`.

And just to be sure, the `flare.exe` (and the DLLs) should be in the same directory that contains the `mods/` directory as it's the easiest way to get the engine to find the mods.

**Comment by TheAJArts** [comment]
`flare.exe --version`

> 0 [main] flare 7409 cygwin_exception::open_stackdumpfile: Dumping stack trace to flare.exe.stackdump

**Comment by dorkster** [comment]
The fact that it's crashing when using the `--version` flag means that neither mods nor SDL are the issue. Not sure what to do here besides suggest building with MinGW instead of Cygwin (which is how I do our Windows builds).

So at this point, use Windows' cmd.exe or [Cmder](https://cmder.net/) if you want something nicer.

There should be something like `mingw64-g++.exe` in the `bin/` directory (you might want to add that directory to your Windows PATH variable). If you copied the SDL files to the MinGW directory like I mentioned earlier, you should be able to run the "g++" command you were running with the `mingw64-g++.exe`.

**Comment by TheAJArts** [comment]
I may just switch over to 32-bit to compile and run with..
Also, mingw is installed as 32-bit as I am just now noticing. All of the commands I've been running have been mainly in cygwin terminal (and ran flare.exe in cmd) cause the mingw32-g++.exe will not load.

I am uninstalling cygwin and wingw32 then I will install wingw32 again to see if I can get the terminal to run

**Comment by TheAJArts** [comment]
Okay I reinstalled mingw32 and still am not able to run 

> mingw32-g++.exe

**Comment by TheAJArts** [comment]
Now I'm in MinGW, all of the libs are in their right locations.
No more errors other than:

`c:/mingw/bin/../lib/gcc/mingw32/8.2.0/../../../libmingwex.a(vsnprintf.o):(.text+0x0): multiple definition of `vsnprintf'; `
```
\AppData\Local\Temp\ccT18aLf.o:Utils.cpp:(.text$vsnprintf[_vsnprintf]+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
```

The file gets deleted when the process finishes.. I've search the pc for vsnprintf

**Comment by TheAJArts** [comment]
Any ideas?

**Comment by dorkster** [comment]
For the `vsnprintf` issue, does adding `#include <stdio.h>` to Utils.cpp help? I didn't run into it when compiling myself, but it may be a result of different environments.

For the DLLs, did you get them from one of the "Runtime Binaries" archives on the SDL site? If not, here are links:

### 32-bit
- https://www.libsdl.org/release/SDL2-2.0.9-win32-x86.zip
- https://www.libsdl.org/projects/SDL_image/release/SDL2_image-2.0.5-win32-x86.zip
- https://www.libsdl.org/projects/SDL_mixer/release/SDL2_mixer-2.0.4-win32-x86.zip
- https://www.libsdl.org/projects/SDL_ttf/release/SDL2_ttf-2.0.15-win32-x86.zip

### 64-bit
- https://www.libsdl.org/release/SDL2-2.0.9-win32-x64.zip
- https://www.libsdl.org/projects/SDL_image/release/SDL2_image-2.0.5-win32-x64.zip
- https://www.libsdl.org/projects/SDL_mixer/release/SDL2_mixer-2.0.4-win32-x64.zip
- https://www.libsdl.org/projects/SDL_ttf/release/SDL2_ttf-2.0.15-win32-x64.zip

**Comment by TheAJArts** [comment]
Okay, so I went ahead and removed all of the libraries that I had in and tried your links, (most of which I did in fact have installed) and replaced with those. Same error.

I read someones comment on google here: https://stackoverflow.com/questions/56861521/how-to-fix-the-procedure-entry-point-sdl-rwclose-could-not-be-located-in-the-dy

They stated to use an earlier version, so I replaced them with the 2.0.4 version and it is now RUNNING!

https://ibb.co/dbCBDPs
